processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -147,31 +387,115 @@ public int indexOf(T value)
     return bufferIndexed.iterator();
   }
 
-  private final ByteBuffer theBuffer;
-  private final ObjectStrategy<T> strategy;
-  private final boolean allowReverseLookup;
-  private final int size;
+  public long getSerializedSize()
+  {
+    if (valueBuffers.size() != 1) {
+      throw new UnsupportedOperationException("Method not supported for version 2 GenericIndexed.");
+    }
+    return theBuffer.remaining() + 2 + 4 + 4;
+  }
 
-  private final int valuesOffset;
-  private final BufferIndexed bufferIndexed;
+  public void writeToChannel(WritableByteChannel channel) throws IOException
+  {
+    //version 2 will always have more than one buffer in valueBuffers.
+    if (valueBuffers.size() == 1) {
+      channel.write(ByteBuffer.wrap(new byte[]{VERSION_ONE, allowReverseLookup ? (byte) 0x1 : (byte) 0x0}));
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(theBuffer.remaining() + 4)));
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(size)));
+      channel.write(theBuffer.asReadOnlyBuffer());
+    } else {
+      throw new UnsupportedOperationException(
+          "GenericIndexed serialization for V2 is unsupported. Use GenericIndexedWriter instead.");
+    }
+  }
 
-  GenericIndexed(
-      ByteBuffer buffer,
-      ObjectStrategy<T> strategy,
-      boolean allowReverseLookup
-  )
+  /**
+   * Create a non-thread-safe Indexed, which may perform better than the underlying Indexed.
+   *
+   * @return a non-thread-safe Indexed
+   */
+  public GenericIndexed<T>.BufferIndexed singleThreaded()
   {
-    this.theBuffer = buffer;
-    this.strategy = strategy;
-    this.allowReverseLookup = allowReverseLookup;
+    final ByteBuffer copyHeaderBuffer = headerBuffer.asReadOnlyBuffer();
```

Comment:
I thought you were referring to copy of data. Yes there is a copy of object and can be avoided . Will update the PR.